shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -541,5 +542,117 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
     typeMethods(typeTypeDefs(ctx.copy(env = allEnv, mthEnv = allMthEnv, tyDefs = allDefs)))
   }
   
-  
+  /**
+    * Finds variances of all type variables in the given type definitions with the given
+    * context using a fixed point computation. The algorithm starts with each type variable
+    * as bivariant and then keeps updating based on the types it encounters.
+    *
+    * @param tyDefs
+    * @param ctx
+    * @return Mapping from type variable to variance info
+    */
+  def computeVariances(tyDefs: List[TypeDef], ctx: Ctx): Unit = {
+    var varianceUpdated: Bool = false;
+
+    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = {
+        updateVariance(fieldTy.ub, curVariance.flip(VarianceInfo.co()))
+        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip(VarianceInfo.contra())))
+    }
+
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = ty match {
+      case ProxyType(underlying) => updateVariance(underlying, curVariance)
+      case TraitTag(_) | ClassTag(_, _) => ()
+      case ExtrType(pol) => ()
+      case t: TypeVariable => val _ = tyDef.tvarVariance.updateWith(t)(tVariance => tVariance match {
+        case None => Some(curVariance)
+        case Some(value) =>
+          val newVariance = value && curVariance
+          if (newVariance =/= curVariance) varianceUpdated = true;
+          Some(newVariance)
+      })
+      case RecordType(fields) => fields.foreach {
+        case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+      }
+      case NegType(negated) =>
+        updateVariance(negated, curVariance.flip(VarianceInfo.contra()))
+      case TypeRef(defn, targs) =>
+        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
+        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
+          val tvarVariance = if (curVariance === VarianceInfo.contra())
+              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
+            else
+              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
+          updateVariance(targ, tvarVariance)
```

Comment:
Actually this case should read:
```suggestion
          val tvarVariance = typeRefDef.tvarVariances.getOrElseUpdate(tvar, VarianceInfo.bi)
          if (tvarVariance.isInvariant) updateVariance(targ, tvarVariance)
          else if (!tvarVariance.isCovariant) updateVariance(targ, curVariance.flip)
          else if (!tvarVariance.isContravariant) updateVariance(targ, curVariance)
```
or equivalently (probably cleaner):
```suggestion
          typeRefDef.tvarVariances.getOrElseUpdate(tvar, VarianceInfo.bi) match {
            case in @ VarianceInfo(false, false) => updateVariance(targ, in)
            case VarianceInfo(true, false) => updateVariance(targ, curVariance)
            case VarianceInfo(false, true) => updateVariance(targ, curVariance.flip)
            case VarianceInfo(true, true) =>
          }
```